### ISSUE TITLE
Creating/destroying units repeatedly using HTTP API is causing units to fail to launch

### ISSUE THREAD

**Issue by bmorton** [coreos/fleet#1131]
This is a lot like #1079 and #997 but I wanted to give another datapoint if it helps as the situation is a little different, but the issue seems to manifest the same.  This is in Vagrant on CoreOS alpha (593.0.0).

I am doing 3 PUTs quickly:

```
PUT /units/hello-world-0fbb804@1.service
PUT /units/hello-world-0fbb804@2.service
PUT /units/hello-world-0fbb804@3.service
```

When I do this, rarely do all units come up.  Usually I see a similar error to those other mentioned issues that looks something like `Failed to trigger systemd unit hello-world-0fbb804@2.service start: Unit hello-world-0fbb804@2.service failed to load: No such file or directory.`

Here's some info from my journal that seems suspect:

```
Feb 20 10:07:59 deployster fleetd[548]: INFO engine.go:272: Scheduled Unit(hello-world-0fbb804@1.service) to Machine(75f8fc8356be438fae39cc3fe46ea7f5)
Feb 20 10:07:59 deployster fleetd[548]: INFO reconciler.go:163: EngineReconciler completed task: {Type: AttemptScheduleUnit, JobName: hello-world-0fbb804@1.service, MachineID: 75f8fc8356be438fae39cc3fe46ea7f5, Reason: "target state launched and unit not scheduled"}
Feb 20 10:07:59 deployster fleetd[548]: INFO engine.go:272: Scheduled Unit(hello-world-0fbb804@2.service) to Machine(75f8fc8356be438fae39cc3fe46ea7f5)
Feb 20 10:07:59 deployster fleetd[548]: INFO reconciler.go:163: EngineReconciler completed task: {Type: AttemptScheduleUnit, JobName: hello-world-0fbb804@2.service, MachineID: 75f8fc8356be438fae39cc3fe46ea7f5, Reason: "target state launched and unit not scheduled"}
Feb 20 10:07:59 deployster fleetd[548]: INFO engine.go:272: Scheduled Unit(hello-world-0fbb804@3.service) to Machine(75f8fc8356be438fae39cc3fe46ea7f5)
Feb 20 10:07:59 deployster fleetd[548]: INFO reconciler.go:163: EngineReconciler completed task: {Type: AttemptScheduleUnit, JobName: hello-world-0fbb804@3.service, MachineID: 75f8fc8356be438fae39cc3fe46ea7f5, Reason: "target state launched and unit not scheduled"}
Feb 20 10:08:02 deployster fleetd[548]: INFO manager.go:262: Writing systemd unit hello-world-0fbb804@1.service (720b)
Feb 20 10:08:02 deployster fleetd[548]: INFO manager.go:198: Instructing systemd to reload units
Feb 20 10:08:02 deployster systemd[1]: Reloading.
...
Feb 20 10:08:03 deployster fleetd[548]: INFO manager.go:262: Writing systemd unit hello-world-0fbb804@2.service (720b)
Feb 20 10:08:03 deployster fleetd[548]: INFO reconcile.go:321: AgentReconciler completed task: type=LoadUnit job=hello-world-0fbb804@1.service reason="unit scheduled here but not loaded"
...
Feb 20 10:08:03 deployster fleetd[548]: INFO manager.go:134: Triggered systemd unit hello-world-0fbb804@1.service start: job=9747
Feb 20 10:08:03 deployster fleetd[548]: INFO reconcile.go:321: AgentReconciler completed task: type=StartUnit job=hello-world-0fbb804@1.service reason="unit currently loaded but desired state is launched"
Feb 20 10:08:03 deployster systemd[1]: Starting hello-world...
Feb 20 10:08:03 deployster systemd[1]: Started Update CA bundle at /etc/ssl/certs/ca-certificates.crt.
Feb 20 10:08:03 deployster fleetd[548]: INFO manager.go:262: Writing systemd unit hello-world-0fbb804@3.service (720b)
Feb 20 10:08:03 deployster fleetd[548]: INFO reconcile.go:321: AgentReconciler completed task: type=LoadUnit job=hello-world-0fbb804@2.service reason="unit scheduled here but not loaded"
Feb 20 10:08:03 deployster fleetd[548]: ERROR manager.go:136: Failed to trigger systemd unit hello-world-0fbb804@2.service start: Unit hello-world-0fbb804@2.service failed to load: No such file or directory.
Feb 20 10:08:03 deployster fleetd[548]: INFO reconcile.go:321: AgentReconciler completed task: type=StartUnit job=hello-world-0fbb804@2.service reason="unit currently loaded but desired state is launched"
Feb 20 10:08:03 deployster fleetd[548]: INFO reconcile.go:321: AgentReconciler completed task: type=LoadUnit job=hello-world-0fbb804@3.service reason="unit scheduled here but not loaded"
Feb 20 10:08:03 deployster fleetd[548]: ERROR manager.go:136: Failed to trigger systemd unit hello-world-0fbb804@3.service start: Unit hello-world-0fbb804@3.service failed to load: No such file or directory.
Feb 20 10:08:03 deployster fleetd[548]: INFO reconcile.go:321: AgentReconciler completed task: type=StartUnit job=hello-world-0fbb804@3.service reason="unit currently loaded but desired state is launched"
```

Its possible that I am using the HTTP API incorrectly to launch 3 units of the same application, so guidance there would be appreciated if there's a better way to do this.

**Comment by bmorton** [comment]
It seems like one of the 3 will launch and the other 2 will fail, but I don't see a pattern for which fail or succeed.  I've tried adding a 5 second sleep in-between the calls thinking that might help, but I see the same thing.

**Comment by bmorton** [comment]
I've tried this on the latest stable as well (557.2.0) and still get the same thing.  Any hints if I'm doing something wrong or if I should keep digging somewhere else?

**Comment by bdehamer** [comment]
Seeing something similar to this issue on 557.2.0. I've got a few units that will fail to start about 50% of the time. I've got a script that demonstrates the issue pretty consistently:

https://gist.github.com/bdehamer/9c5303f7ef9d463a9134

This script goes into a loop and simply destroys, submits, loads, and starts the units over-and-over (you need the Fleet API bound to TCP port 49153 for the script to work). Things usually run fine the first time through, but will fail on the second execution with a not-found error:

```
core@core-01 ~/dev $ fleetctl status DB.service
● DB.service
   Loaded: not-found (Reason: No such file or directory)
   Active: inactive (dead) since Mon 2015-02-23 16:29:25 UTC; 10s ago
 Main PID: 5605 (code=exited, status=0/SUCCESS)
```

**Comment by bmorton** [comment]
I'm glad to hear I'm not the only one having this issue.  I do have some more notes on when this fails and a workaround I've been using.  I've been doing a lot of testing of switching between two versions of an application and making sure I can bring up a new version with zero downtime.  I'm doing it with the API, but I've replicated what I'm doing by using `fleetctl` and this pretty consistently triggers the issue:

```Shell
$ fleetctl start hello-world-ee0f771@{1..3}.service
# all units start properly, only ee0f771 running
$ fleetctl start hello-world-0fbb804@{1..3}.service
# all units start properly, ee0f771 and 0fbb804 running
$ fleetctl destroy hello-world-ee0f771@{1..3}.service
# all units destroy properly, only 0fbb804 running
$ fleetctl start hello-world-ee0f771@{1..3}.service
# some units fail to start (usually 1 works, 2 fail)
```

In this example, I'm switching between two versions of `hello-world` (`ee0f771` and `0fbb804`) and using a naming convention to keep the unit names straight.  The failure that occurs is the same as the log above (fleetd triggers a systemd reload after 1 unit file is loaded and then the other 2 fail to be found).

If I switch and do this, I don't ever see any units fail to load (making sure to rename the unit file on disk between the changes):

```Shell
$ fleetctl start hello-world-ee0f771-20140225-0750@{1..3}.service
# all units start properly, only ee0f771 running
$ fleetctl start hello-world-0fbb804-20140225-0751@{1..3}.service
# all units start properly, ee0f771 and 0fbb804 running
$ fleetctl destroy hello-world-ee0f771-20140225-0750@{1..3}.service
# all units destroy properly, only 0fbb804 running
$ fleetctl start hello-world-ee0f771-20140225-0752@{1..3}.service
# some units fail to start (usually 1 works, 2 fail)
```

Basically, if I don't use the same name over again and introduce a timestamp to the end of the unit name, I don't ever see any units fail to load.

Is this expected and should I be doing this a different way?  [Here's a gist](https://gist.github.com/bmorton/343735730aa8400b9f3f) with my unit files that can be used to reproduce this.  Again, I'm using Vagrant with latest alpha and just one node.

**Comment by bmorton** [comment]
Oops, my second (correct) example had a small copy/paste error.  I've edited the above message and its correct now.

**Comment by pirelenito** [comment]
I can confirm this same issue on CoreOS stable (557.2.0)

**Comment by pirelenito** [comment]
Thanks @bmorton for your timestamp workaround.

I've been using it in my environment and the issues disappeared.

**Comment by mpchlets** [comment]
I am seeing the same issue, a timestamp does work around the problem, however when I look at `systemctl list-unit-files`, I still see the old files hanging around:

    core-data@1425666060240.service                                               enabled 
    core-data@1425666313949.service                                               enabled 
    core-data@1425666438330.service                                               enabled 
    core-data@1425666482205.service                                               enabled 
    core-data@1425666610816.service                                               enabled 
    core-data@1425666649896.service                                               enabled 
    core-data@1425666755216.service                                               enabled

Fleet is not properly removing the files from the system, so if they are the same name, it causes the issue we are seeing.

**Comment by ejholmes** [comment]
@bmorton setting `KillMode=none` seems to have solved this for me.

**Comment by patrickbcullen** [comment]
I am seeing the same issue. Every other submit/start/destroy cycle fails with the No such file or directory error. I was able to narrow it down to fleet 0.9.0. Fleet 0.8.3 does not have this problem. This issue is probably the same thing https://github.com/coreos/fleet/issues/1144

**Comment by tleyden** [comment]
I just filed #1149 which may be a duplicate of this and #1137.  #1149 hopefully adds some value with steps to reproduce.

**Comment by bcwaldon** [comment]
@bmorton the fact that you are using the fleet API is likely a red herring. I suspect this is the same as #1158

**Comment by bmorton** [comment]
:+1: I'll track that issue instead.  Thanks!

**Comment by pateld982** [comment]
I'm also running into same issue while creating and starting fleet units VIA fleet API. Is there a work around for this issue? 

Looks like fleet is only able to start three units. If I remove one unit it will pick up the next unit and start. See below for out put for fleetctl logs and systemctl .

$ fleetctl list-unit-files | grep 'twitter_nsq_producer_test'
producer_test@2d63be2e_fdae_484c_bca3_b62ffc92da31.service          ade9947 launched        launched        d422662c.../0.108.69.166
producer_test@2e14e481_9ac9_4ec6_b531_6e94dc0b2c66.service          ade9947 launched        inactive        -
producer_test@2f2032e4_8012_4bd6_a990_a1de0a74f2cb.service          ade9947 launched        inactive        -
producer_test@aae5da63_0515_4eb8_8c35_7fd48727887e.service          ade9947 launched        launched        s3f6338c.../0.108.69.168
producer_test@def77948_5e4a_4f62_a5e5_159905228e42.service          ade9947 launched        inactive        -
producer_test@f4cbb3e9_6092_4c21_a357_53261fef8f1a.service          ade9947 launched        launched        1a42863e.../0.108.69.164

$ systemctl list-unit-files | grep 'twitter_nsq_producer_test'
producer_test@21119efe_178e_464e_89d9_edb4ef50ea83.service  enabled
producer_test@56d0c338_da5c_4b1f_b2b2_af4ddba21f64.service  enabled
producer_test@58b49b39_9df1_4c6b_b8bb_c7f8641a8987.service  enabled
producer_test@5ca785b9_3743_4c9d_a2cf_d0560b5ae5dd.service  enabled
producer_test@f4cbb3e9_6092_4c21_a357_53261fef8f1a.service  enabled
